sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMemoryLeaseManager.scala

```diff
@@ -0,0 +1,347 @@
+   *                       nonBlocking = true)
+   * @param nonBlocking can the request block to get the requested resources, or not.
+   * @return the lease.
+   */
+  def requestLease(tc: TaskContext,
+      optionalAmount: Long,
+      requiredAmount: Long,
+      nonBlocking: Boolean): MemoryLease = synchronized {
+    val taskAttemptId = tc.taskAttemptId()
+    if (nonBlocking && requiredAmount > 0) {
+      throw new IllegalArgumentException("Non-blocking requests with a required minimum amount " +
+          "of memory are not supported, because they are too likely to fail randomly")
+    }
+    val alreadyRequested = getTotalLease(tc)
+    if (requiredAmount + alreadyRequested > memoryForLease) {
+      throw new IllegalArgumentException(s"Task: $taskAttemptId requested at least " +
+          s"$requiredAmount more bytes, but already has leased $alreadyRequested bytes which " +
+          s"would got over the total for the worker of $memoryForLease bytes.")
+    }
+
+    if (nonBlocking && (!taskPriorityQueue.isEmpty || memoryAvailable <= 0)) {
+      // This would block so just indicate it right now.
+      return new MemoryLease {
+        override def leaseAmount: Long = 0
+
+        override def close(): Unit = {}
+      }
+    }
+
+    val totalAmountRequested = requiredAmount + optionalAmount
+    val amountToRequest = if (nonBlocking && totalAmountRequested > memoryAvailable) {
+      logDebug(s"Task: $taskAttemptId requested $totalAmountRequested bytes in a " +
+          s"non-blocking request, but only $memoryAvailable is available. The request will " +
+          s"be reduced to $memoryAvailable.")
+      memoryAvailable
+    } else if (totalAmountRequested + alreadyRequested > memoryForLease) {
+      logWarning(s"Task: $taskAttemptId requested $totalAmountRequested bytes, but has " +
+          s"already requested $alreadyRequested bytes. This would go over the total for the " +
+          s"worker $memoryForLease reducing the request on the hope that this was an " +
+          s"overestimate.")
+      memoryForLease - alreadyRequested
+    } else {
+      totalAmountRequested
+    }
+
+    val task = activeTasks.computeIfAbsent(taskAttemptId, _ => new TrackingForTask)
+
+    if (!taskPriorityQueue.isEmpty || memoryAvailable < amountToRequest) {
+      // Need to block until there is memory available
+      if (task.waitingThreadCount > 0 || task.canWakeUp) {
+        throw new IllegalStateException("MULTIPLE THREADS TRYING TO REQUEST RESOURCES FOR " +
+            "A SINGLE TASK")
+      }
+      // We should never hit this, because the checks before should avoid blocking, but just to be
+      // safe we are going to check.
+      assert(!nonBlocking)
+
+      // Release all of the currently used memory so it can be reused by another task if needed.
+      memoryAvailable += task.totalUsed
+      wakeTasksThatCanRun()
+
+      // Update the task request state for what is now needed
+      task.waitingThreadCount = 1
```

Comment:
I'm a bit confused by this, as it indicates `waitingThreadCount` shouldn't be an int.  It seems like the value can never be anything except 0 or 1, implying it's just a boolean.  That aligns with the earlier assert that multiple threads cannot be asking for resources for the same task, therefore we should never see multiple threads waiting for a particular task.